### PR TITLE
Use lambda instead of AZStd::bind

### PR DESCRIPTION
**Pull request by nemerle**
Since the engine target version is c++17 using std::bind is a likely indicator of older code :)

Signed-off-by: nemerle <96597+nemerle@users.noreply.github.com>

### CODE REVIEW

**Review by gadams3 — APPROVED** [review]

File: Code/Tools/AssetProcessor/native/AssetManager/SourceFileRelocator.cpp
Comment (AMZN-stankowi) [inline Code/Tools/AssetProcessor/native/AssetManager/SourceFileRelocator.cpp:1029]: Have you tested this with Perforce integration? This change makes me nervous because I don't think anyone's regularly testing Perforce integration with the engine in git now, and this change directly touches Perforce integration code.

Comment (rgba16f) [reply Code/Tools/AssetProcessor/native/AssetManager/SourceFileRelocator.cpp:1029]: Is perforce integration a feature we want to maintain?

Comment (nemerle) [reply Code/Tools/AssetProcessor/native/AssetManager/SourceFileRelocator.cpp:1029]: I can't see how converting bind to a lambda could impact perforce integration at all, especially since SourceControlResponseCallbackBulk is `AZStd::function` which should consume any callable ?

Comment (AMZN-stankowi) [reply Code/Tools/AssetProcessor/native/AssetManager/SourceFileRelocator.cpp:1029]: > Is perforce integration a feature we want to maintain?
Even if we're using git now, Perforce is standard in the games industry, and last I heard it's still supported for our tools.

> I can't see how converting bind to a lambda could impact perforce integration at all, especially since SourceControlResponseCallbackBulk is AZStd::function which should consume any callable ?
I don't expect the change to impact it, but we've been burned by that before with seemingly safe changes breaking things.

Comment (nemerle) [reply Code/Tools/AssetProcessor/native/AssetManager/SourceFileRelocator.cpp:1029]: Is there some kind of test I could run to verify this, without buying Perforce?

Comment (AMZN-stankowi) [reply Code/Tools/AssetProcessor/native/AssetManager/SourceFileRelocator.cpp:1029]: Not that I'm aware of, I think all Perforce testing is currently manual right now, and is pretty time consuming. I'll click approve for now, but I'm still nervous about this changing this logic due to the infrequency of testing in this area.

**Review by AMZN-stankowi — APPROVED** [review]

File: Code/Tools/AssetProcessor/native/AssetManager/SourceFileRelocator.cpp
Comment (nemerle) [reply Code/Tools/AssetProcessor/native/AssetManager/SourceFileRelocator.cpp:1029]: The logic shouldn't have changed at all. Simple replacement of older library construct with a language one.

Comment (hultonha) [reply Code/Tools/AssetProcessor/native/AssetManager/SourceFileRelocator.cpp:1029]: Calling @lumberyard-employee-dm who can probably sanity check this for correctness (and will be happy to see an `AZStd::bind` usage removed 🙂)

**Review by lawsonamzn — APPROVED** [review]
This is fine, get rid of bind.

**Comment by lawsonamzn** [comment]
I've kicked off a build at  https://jenkins.build.o3de.org/job/O3DE/view/change-requests/job/PR-2658/1/ 
if it succeeds, I'll go ahead and merge this in.  Lambda all the binds!

**Comment by nemerle** [comment]
Don't tempt me @lawsonamzn, or I might just go and do it :D

File: Code/Tools/AssetProcessor/native/AssetManager/SourceFileRelocator.cpp
Comment (lumberyard-employee-dm) [reply Code/Tools/AssetProcessor/native/AssetManager/SourceFileRelocator.cpp:1029]: To be truly safe, what should occur is to not use `&` for capturing the local variable parameters by reference, but to capture each argument as part of the lambda capture.

```
[&relocationContainer, &waitSignal, &errorCount, ](bool success, AZStd::vector<SourceControlFileInfo> info)
```

Comment (nemerle) [reply Code/Tools/AssetProcessor/native/AssetManager/SourceFileRelocator.cpp:1029]: To be honest, I've followed the most common way lambda is used around the codebase, and in this case, capturing by`&` should be semantically the same?

Comment (lumberyard-employee-dm) [reply Code/Tools/AssetProcessor/native/AssetManager/SourceFileRelocator.cpp:1029]: It should be, but this is to prevent "over" capture of local variables

Comment (nemerle) [reply Code/Tools/AssetProcessor/native/AssetManager/SourceFileRelocator.cpp:1029]: I agree that 'blanket' `&` in a lambda's capture list can cause problems, but, in this case, the lambda's body is quite small, not like for example the lambda in `Remapper<IdType, AllowDuplicates>::RemapIds` :smile: 

So while I can easily rewrite it as suggested, there are ~600 instances of `[&]` capture lambdas around the codebase that will likely need similar treatment?

Comment (lumberyard-employee-dm) [reply Code/Tools/AssetProcessor/native/AssetManager/SourceFileRelocator.cpp:1029]: We should keep code that is working as is.
So we shouldn't proactively go out of the way to have "lambdas only capture what they use".
But over capture can cause unnecessary memory allocations when stored in containers like a std::function, which contains a small buffer that can store captured function objects within a small buffer internally.

Comment (nemerle) [reply Code/Tools/AssetProcessor/native/AssetManager/SourceFileRelocator.cpp:1029]: Oh, I fully agree and understand that keeping things lean by design is a good goal. I'll update that lambda capture tomorrow, to contain only used variables, and would like to suggest updating the https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md to include this guidance. (@hultonha, @lawsonamzn  ? )

Comment (lumberyard-employee-dm) [reply Code/Tools/AssetProcessor/native/AssetManager/SourceFileRelocator.cpp:1029]: The `&` capture isn't blocking this change from going in.
I have no issues with it as it stands and because the change has made through AR and has enough approvals I am going to merge it to the development.

Comment (hultonha) [reply Code/Tools/AssetProcessor/native/AssetManager/SourceFileRelocator.cpp:1029]: > 
> 
> Oh, I fully agree and understand that keeping things lean by design is a good goal. I'll update that lambda capture tomorrow, to contain only used variables, and would like to suggest updating the https://github.com/o3de/sig-core/blob/main/governance/Coding-Standards-and-Style-Guide.md to include this guidance. (@hultonha, @lawsonamzn ? )

That's a great idea @nemerle, might be worth seeing if there's somewhere for that recommendation to go if we have a section on lambdas (I think the other thing too would be to bring this up in the sig-core Discord to discuss - you can now start a thread about it there which is cool! 😄)

**Pull request merged**
